### PR TITLE
Fixed-point quantize transpiler.

### PR DESCRIPTION
**Pull request by qingqing01**

### CODE REVIEW

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+    else:
+        return var_name
+
+
+def _is_float(v):
+    return isinstance(v, float) or isinstance(v, np.float32)
+
+
+def quant(x, scale, num_bits):
+    y = np.round(x / scale * ((1 << (num_bits - 1)) - 1))
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/quantize/quantize_transpiler.py:77]: Maybe check scale larger than epsilon

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+            activation_bits (int): quantization bit number for activation.
+            activation_quantize_type (str): quantization type for activation,
+                now support 'abs_max', 'range_abs_max'. If use 'abs_max' mode,
+                the quantization scale will be calculated dynamically each step
+                in both training and testing period. If use 'range_abs_max',
+                a static quantization scale will be calculated during training
+                and used in inference.
+            weight_quantize_type (str): quantization type for weights,
+                support 'abs_max'. The 'range_abs_max' usually is not used for
+                weight, since weights are fixed once the model is well trained.
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/quantize/quantize_transpiler.py:104]: So, is range_abs_max static or not? From the previous comment, range_abs_max sounds like static.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+                    quant_type = self.weight_quantize_type if var.name \
+                        in params else self.activation_quantize_type
+
+                    quant_var, scale_var = self._insert_quant_op(
+                        block, idx, var, quant_bits, quant_type)
+                    dequant_var = self._insert_dequant_op(
+                        block, idx + 1, quant_var, scale_var, quant_bits)
+                    dequanted_vars[block_id][name] = dequant_var
+                # rename the forward op inputs
+                op.rename_input(name, dequant_var.name)
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/quantize/quantize_transpiler.py]: why do you rename this op's input to dequant_var.name? I think dequant_var will affect next op's input?

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+                if name in dequanted_vars[block_id]:
+                    dequant_var = dequanted_vars[block_id][name]
+                    op.rename_input(name, dequant_var.name)
+                    no_dequanted_input_vars = False
+            if no_dequanted_input_vars:
+                raise ValueError("There is no dequanted inputs for op %s." %
+                                 (op.type))
+
+        with program_guard(program, startup_program):
+            self._create_globael_step()
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/quantize/quantize_transpiler.py]: global?

**Comment by qingqing01** [comment]
```
[14:47:24][Step 1/1] + echo 'current pr 13319 got approvals: FALSE'
[14:47:24][Step 1/1] + '[' FALSE == FALSE ']'
[14:47:24][Step 1/1] + echo 'You must have at least 2 approvals for the api change!'
[14:47:24][Step 1/1] + exit 1
[14:47:24][Step 1/1] current pr 13319 got approvals: FALSE
[14:47:24][Step 1/1] You must have at least 2 approvals for the api change!
```

@panyx0718 @reyoung Please help to review.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+                    if name in op_out_rename_map[block_id]:
+                        op.rename_input(name, op_out_rename_map[block_id][name])
+
+                if op_type in self.fake_quant_op_types:
+                    in_arg_name = op.input('X')[0]
+                    if in_arg_name in persistable_vars:
+                        if self.weight_quantize_type == 'abs_max':
+                            param = _load_var(in_arg_name)
+                            scale_v = np.max(np.abs(param))
+                        else:
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/quantize/quantize_transpiler.py:324]: perhaps elif?

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+    def _remove_unused_var(self, program):
+        for block in program.blocks:
+            args = []
+            for op in block.ops:
+                args += op.input_arg_names
+                args += op.output_arg_names
+            args = list(set(args))
+            for var in block.vars.keys():
+                if var not in args:
+                    block._remove_var(var)
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/quantize/quantize_transpiler.py]: does this remove var in scope as well?

**Review by panyx0718 — DISMISSED** [review]

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 1 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py:77]: This function is used to save weight as int8 range.  The scale is the max_abs of weights. It must be larget than 0. If 0, the weights is bad. So, no need to add epsilon.

And usually, it needs to clip data out of  int8 range, but since this function is used for weights, the abs max is the true abs max, the quanted data does not exceed the range.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 2 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py:104]: Yeah, range_abs_max is called static quantization. Like batch norm, it will get a `max_abs` scale in training for activation in inference. So, there is no need to calculate `max_abs` dynamically in inference for each input data.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 3 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py]: x -> quant ->fc/conv -> dequant

Since all these are  linear calculation, it is equivalent to:

x -> quant ->dequant -> fc/conv.

In our implementation, we use `x -> quant ->dequant -> fc/conv.` So dequant_var is the input of this op.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 4 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py]: Oh. Thanks, I will rename this function name.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 6 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py]: This is in compile time,  the var is newed in scope in runtime.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 5 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py:324]: Since now only `abs_max` is dynamic quantization, needs to calculate max_bas value. `else` is ok now.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 4 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py]: Done.

**Review by reyoung — DISMISSED** [review]
Looks good for API changes.

File: python/paddle/fluid/contrib/tests/test_quantize_transpiler.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+                    if not is_test:
+                        opt = fluid.optimizer.Adam(learning_rate=0.001)
+                        opt.minimize(loss)
+            return [img, label], loss
+
+        main = fluid.Program()
+        startup = fluid.Program()
+        test_program = fluid.Program()
+
+        feeds, loss = build_program(main, startup, True)
```
Comment (hjchen2) [inline python/paddle/fluid/contrib/tests/test_quantize_transpiler.py]: Why the is_test option is true here? I think here you want to build a train program.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+        Args:
+            program (Program): the input program to be transpile.
+        """
+
+        self.is_test = True
+        scope = global_scope() if scope is None else scope
+        program = default_main_program() if program is None else program
+
+        if fuse_bn:
+            bn_fuse_transpiler = BNFuseTranspiler()
```
Comment (hjchen2) [inline python/paddle/fluid/contrib/quantize/quantize_transpiler.py:230]: Weights are trained with int8, otherwise bn scale and beta are trained with float. Fuse bn directly in inference will maybe cause precision loss.

File: python/paddle/fluid/contrib/tests/test_quantize_transpiler.py (hunk 7 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/tests/test_quantize_transpiler.py]: Yeah, Thanks very much.

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 8 above)
Comment (qingqing01) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py:230]: Yeah,  here is a choice to do test.

**Review by panyx0718 — DISMISSED** [review]

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 1 above)
Comment (panyx0718) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py:77]: what if it's some bias that the scale is very small? someone will have a hard time debugging

File: python/paddle/fluid/contrib/quantize/quantize_transpiler.py (hunk 3 above)
Comment (panyx0718) [reply python/paddle/fluid/contrib/quantize/quantize_transpiler.py]: what's the point of `quant ->dequant`?

**Review by panyx0718 — APPROVED** [review]

**Review by reyoung — APPROVED** [review]
